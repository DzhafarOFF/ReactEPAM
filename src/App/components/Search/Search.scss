.search{
    padding: 20px;
    margin: 0 auto;
    display: flex;
    flex-direction: column;
    max-width: 1140px;
    color: #fefefe;
}

.search-container {
    background-image: url(./netflixroulette.jpg);
    background-size: cover;
    background-color: #404040;
    background-blend-mode: overlay;
}

.search-title{
    color:lighten($color: #E50914, $amount: 20) ;
}

.search-filter-container {
    width: 100%;
    display: flex;
    justify-content: space-between;
}