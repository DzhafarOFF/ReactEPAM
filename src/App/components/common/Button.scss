.button{
    padding: 10px;
    margin: 10px;
    background-color: #E54B4B;
    color: #fefefe;
    border: 0;
    border-radius: 5%;
    text-transform: uppercase;
    font-weight: bolder;
    transition: 0.2s;
    cursor: pointer;
}
button:hover {
 background-color: lighten($color: #E54B4B, $amount: 5);
}
.button:disabled {
    background-color: grey;
}