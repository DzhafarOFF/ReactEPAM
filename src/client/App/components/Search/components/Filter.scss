.filter-title{
    display: inline-block;
}
.filter-option{
    display: inline-block;
}