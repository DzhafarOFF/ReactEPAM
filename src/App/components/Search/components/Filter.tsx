import React from 'react';
import Button from '../../common/Button'

type FilterProps = {
    options: string[];
    activeOption: string;
    handleSwitch: (e: React.MouseEvent) => void;
}
const Filter = (props: FilterProps) => {
  const setFilterOptions = props.options.map(
    (option, index) => {
      if (option === props.activeOption) return (<Button
        key = {index}
        content = {option}
        action = {props.handleSwitch}
        disabled = {true}
      />);
      return (<Button
        key = {index}
        content = {option}
        action = {props.handleSwitch}
      />);
    }
  );
  return (
      <>
        <h3>Sort by</h3>
        <div>  
          {setFilterOptions}
        </div>
      </>
  );
};

export default Filter;
