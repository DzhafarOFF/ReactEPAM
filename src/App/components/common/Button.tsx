import React from 'react';
import './Button.scss';

type ButtonsOptions = {
    content: string;
    action: (e: React.MouseEvent) => void;
    value?: string;
    disabled?: boolean;
}

const Button = (props: ButtonsOptions) => {
    return (
        <button
            className='button'
            onClick={props.action}
            value={props.content}
            disabled={props.disabled}>{props.content}
        </button>
    )
}

export default Button;