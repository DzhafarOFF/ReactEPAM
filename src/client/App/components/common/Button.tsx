import './Button.scss';
import React from 'react';

type Props = {
	content: string
	action: (e: React.MouseEvent) => void
	value?: string
	disabled?: boolean
};

const Button: React.FC<Props> = (props: Props) =>
	<button
		className='button'
		onClick={props.action}
		value={props.content}
		disabled={props.disabled}>{props.content}
	</button>
;


export default Button;
