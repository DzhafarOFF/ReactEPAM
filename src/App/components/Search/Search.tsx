import React from 'react';
import Button from '../common/Button';
import Input from '../common/Input';
import Filter from './components/Filter';
import './Search.scss';
import { AppState, SearchQuery, Movie } from '../../typings/types';
import { Dispatch, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { setSearchInput, setSearchFilter } from '../../actions/setSearchQuery';
import { Link } from 'react-router-dom';
import { history } from '../../ConfigureStore';
import { getMovies } from '../../thunkAction/getMovies';
type Props = {
    filterOptions: string[];
    searchQuery: SearchQuery;
    setSearchInput: Function;
    setSearchFilter: Function;
    getMovies: Function;
    fetchedMovies: Movie[];
}

type State = {
    inputValue: string;
    filterOptions: string[];
    activeOption: string;
}
class Search extends React.Component<Props, State>{
    constructor(props: Props) {
        super(props)
        this.state = {
            inputValue: '',
            filterOptions: props.filterOptions,
            activeOption: props.filterOptions[0]
        }
    }

    componentDidUpdate(prevProps: Props) {
        if (prevProps === this.props) {
            this.props.setSearchInput(this.state.inputValue);
            this.props.setSearchFilter(this.state.activeOption)
        }
    }

    componentDidMount() {
        //this.props.setSearchFilter(this.state.activeOption)
        //console.log(this.props.searchQuery.filterOption)
        //console.log(localStorage.getItem('lastSearch'))
        // if (localStorage.getItem('lastSearch')) {
        //     const url = `https://reactjs-cdp.herokuapp.com/movies?search=${localStorage.getItem('lastSearch')}&searchBy=title`;
        //     this.props.getMovies(url)
        //     history.push(`/search/${localStorage.getItem('lastSearch')}`);
        //     localStorage.clear()
        // }
        // else {
        //     const lastPathFragment = history.location.pathname.substring(history.location.pathname.lastIndexOf('/') + 1).trim()
        //     localStorage.setItem('lastSearch', lastPathFragment);
        //     const url = `https://reactjs-cdp.herokuapp.com/movies?search=${lastPathFragment}&searchBy=title`;
        //     this.props.getMovies(url)
        //     history.push(`/search/${lastPathFragment}`);
        // }
        const searchPathFragment = /search\/(.*)/.exec(history.location.pathname);
        
        if (searchPathFragment) {
            localStorage.setItem('lastSearch', searchPathFragment[1]);
            const url = `https://reactjs-cdp.herokuapp.com/movies?search=${searchPathFragment[1]}&searchBy=${this.state.activeOption}`;
            this.props.getMovies(url);
            this.props.setSearchInput(searchPathFragment[1]);
            history.push(`/search/${searchPathFragment[1]}`);
        }
        else {
            const url = `https://reactjs-cdp.herokuapp.com/movies?search=${localStorage.getItem('lastSearch')}&searchBy=${this.state.activeOption}`;
            this.props.getMovies(url)
            history.push(`/search/${localStorage.getItem('lastSearch')}`);
            // console.log('storage clear')
            // localStorage.clear()
        }
        //localStorage.clear()
    }

    handleButtonSubmit = () => {
        this.handleSearch();
        localStorage.setItem('lastSearch', this.state.inputValue);
        history.push(`/search/${this.state.inputValue}`);
    }

    handleInputChange = (e: React.ChangeEvent) => {
        const inputText: string = (e.target as HTMLInputElement).value;
        this.setState({inputValue: inputText});
    }

    handleSearch = () => {
        const url =
        `https://reactjs-cdp.herokuapp.com/movies?search=${this.props.searchQuery.input}&searchBy=${this.props.searchQuery.filterOption}`;
        this.props.getMovies(url)
    }

    handleInputSubmit = (e: React.KeyboardEvent) => {
        if (e.key === 'Enter') {
            this.handleSearch();
            localStorage.setItem('lastSearch', this.state.inputValue);
            history.push(`/search/${this.state.inputValue}`);
        }
    }

    handleSwitchFilter = (e: React.MouseEvent) => {
        this.setState({activeOption: (e.target as HTMLInputElement).value});
    }

    render() {
        return (
            <div className = 'search-container'>
                <div className = 'search'>
                    <Link to='/' className = 'search-title'>
                        <h1 >Netflixroulette</h1>
                    </Link>
                    <h2>Find your movie</h2>
                    <Input
                        type = 'text'
                        changeAction = {this.handleInputChange}
                        submitAction = {this.handleInputSubmit}
                    />
                    <div className = 'search-filter-container'>
                        <Filter
                            options = {this.state.filterOptions}
                            activeOption = {this.state.activeOption}
                            handleSwitch = {
                                (e: React.MouseEvent) => this.handleSwitchFilter(e)
                            }
                        />
                        <Button content='Search' action = {this.handleButtonSubmit}/>
                    </div>
                </div>
            </div>
        )
    }
}

function mapStateToProps (state: AppState){
    return {
        searchQuery: state.searchQuery,
        fetchedMovies: state.fetchedMovies,
    }
}

function mapDispatchToProps (dispatch: Dispatch) {
    return {
        setSearchInput: bindActionCreators(setSearchInput, dispatch),
        setSearchFilter: bindActionCreators(setSearchFilter, dispatch),
        getMovies: bindActionCreators(getMovies, dispatch)
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Search);