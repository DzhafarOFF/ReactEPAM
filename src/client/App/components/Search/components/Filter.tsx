import './Filter.scss';
import Button from '../../common/Button';
import React from 'react';

type Props = {
	options: string[]
	activeOption: string
	handleSwitch: (e: React.MouseEvent) => void
};
const Filter: React.FC<Props> = (props: Props) => {
	const setFilterOptions = props.options.map(
		(option, index) => {
			if (option === props.activeOption) {
				return <Button
					key = {index}
					content = {option}
					action = {props.handleSwitch}
					disabled = {true}
				/>;
			}
			return <Button
				key = {index}
				content = {option}
				action = {props.handleSwitch}
				disabled = {false}
			/>;
		}
	);
	return (
		<div className = 'filter-option'>
			<h3 className='filter-title'>{'Search by'}</h3>
			{setFilterOptions}
		</div>
	);
};

export default Filter;
