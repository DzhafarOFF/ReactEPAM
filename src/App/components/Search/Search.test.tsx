import React from 'react';
import { shallow, render, mount } from 'enzyme';
import Search from './Search';
import Button from '../common/Button';
import Filter from './components/Filter';

const props = {
    filterOptions: ['title', 'genres'],
}

describe('<Search/> rendering', () => {
    const search = shallow(<Search {...props}/>)
    it('should renders correctly',() => {
        expect(search).toMatchSnapshot();
    })
})
describe('<Search/> interacting', () => {
    const search = mount(<Search {...props}/>);

    it('should handle switch Filter event',() => {
        const unactiveOption = search.find(Filter).find(Button).find('button').filterWhere(el => el.prop('disabled') === false);
        const methods = search.instance() as Search;
        jest.spyOn(methods, 'handleSwitchFilter');
        unactiveOption.simulate('click');
        expect(methods.handleSwitchFilter).toBeCalled();
    });
})