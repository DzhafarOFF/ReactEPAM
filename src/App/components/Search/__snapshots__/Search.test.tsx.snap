// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search/> rendering should renders correctly 1`] = `
<div
  className="search-container"
>
  <div
    className="search"
  >
    <h1
      className="search-title"
    >
      Netflixroulette
    </h1>
    <h2>
      Find your movie
    </h2>
    <Input
      changeAction={[Function]}
      submitAction={[Function]}
      type="text"
    />
    <div
      className="search-filter-container"
    >
      <Filter
        activeOption="title"
        handleSwitch={[Function]}
        options={
          Array [
            "title",
            "genres",
          ]
        }
      />
      <Button
        action={[Function]}
        content="Search"
      />
    </div>
  </div>
</div>
`;
